File: packages/ios-swift/Sources/OnChainUX/WalletConnect/WCSessionManager.swift
/**
 * WCSessionManager — Session persistence & auto-reconnect for WalletConnect v2 (iOS).
 *
 * Provides:
 * - UserDefaults-backed session persistence (topic, accounts, chainId)
 * - Session TTL enforcement (default 7 days)
 * - Automatic reconnection with exponential backoff on app foreground
 * - Reliability tracking and reconnect attempt limits
 *
 * Used internally by WCClient; also usable standalone for custom integrations.
 */

import Foundation

// MARK: - Persisted Session Model

/// Codable snapshot of an active WC v2 session for persistence.
struct PersistedSession: Codable {
    let topic: String
    let accounts: [String]
    let chainId: Int
    let peerName: String?
    let createdAt: TimeInterval
    let expiresAt: TimeInterval

    /// Whether this session has exceeded its TTL.
    var isExpired: Bool {
        Date().timeIntervalSince1970 >= expiresAt
    }
}

// MARK: - UserDefaults Keys

enum WCSessionKeys {
    static let sessionData = "wc_v2_persisted_session_v1"
    static let relayUrl = "wc_v2_relay_url"
}

// MARK: - Session Manager

/// Manages WC v2 session lifecycle: persist, restore, expiry-check, reconnect.
public final class WCSessionManager: ObservableObject {

    /// Shared instance for app-wide access.
    public static let shared = WCSessionManager()

    /// UserDefaults backing store (injectable for testing).
    public let defaults: UserDefaults

    /// Default session TTL: 7 days.
    public var sessionTTL: TimeInterval = 7 * 24 * 60 * 60

    /// Whether auto-reconnect is enabled.
    @Published public var autoReconnectEnabled: Bool = true

    /// Maximum reconnect attempts before giving up.
    public var maxReconnectAttempts: Int = 5

    /// Base delay for exponential backoff (seconds).
    public var reconnectBaseDelay: TimeInterval = 2.0

    /// Current reconnect attempt count.
    @Published public private(set) var reconnectAttempts: Int = 0

    /// Pending reconnect work item (for cancellation).
    private var reconnectWorkItem: DispatchWorkItem?

    /// Expiry check timer.
    private var expiryTimer: Timer?

    /// Relay health status.
    @Published public private(set) var relayHealth: RelayHealth = .disconnected

    private init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    // MARK: - Persistence

    /// Save a session snapshot for later restoration.
    public func persistSession(
        topic: String,
        accounts: [String],
        chainId: Int,
        peerName: String? = nil
    ) {
        let session = PersistedSession(
            topic: topic,
            accounts: accounts,
            chainId: chainId,
            peerName: peerName,
            createdAt: Date().timeIntervalSince1970,
            expiresAt: Date().timeIntervalSince1970 + sessionTTL
        )

        if let encoded = try? JSONEncoder().encode(session) {
            defaults.set(encoded, forKey: WCSessionKeys.sessionData)
            defaults.synchronize()
        }
    }

    /// Load the persisted session, if any and not expired.
    public func loadSession() -> PersistedSession? {
        guard let data = defaults.data(forKey: WCSessionKeys.sessionData) else { return nil }
        let session = try? JSONDecoder().decode(PersistedSession.self, from: data)
        // Return nil if expired so caller knows to clean up
        return session?.isExpired == true ? nil : session
    }

    /// Clear persisted session data.
    public func clearSession() {
        defaults.removeObject(forKey: WCSessionKeys.sessionData)
        defaults.synchronize()
    }

    // MARK: - Expiry Monitoring

    /// Start periodic expiry checks (every 60s).
    public func startExpiryMonitoring(onExpired: @escaping () -> Void) {
        stopExpiryMonitoring()
        // Immediate check
        checkExpiry(onExpired: onExpired)
        // Periodic check
        expiryTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
            self?.checkExpiry(onExpired: onExpired)
        }
    }

    /// Stop expiry monitoring.
    public func stopExpiryMonitoring() {
        expiryTimer?.invalidate()
        expiryTimer = nil
    }

    /// Check if the persisted session has expired.
    private func checkExpiry(onExpired: @escaping () -> Void) {
        guard let session = loadSession(), session.isExpired else { return }
        clearSession()
        onExpired()
    }

    // MARK: - Auto-Reconnect

    /// Schedule an auto-reconnect with exponential backoff.
    /// Returns `true` if a reconnect was scheduled.
    @discardableResult
    public func scheduleReconnect(
        attempt: @escaping () async -> Bool
    ) -> Bool {
        guard autoReconnectEnabled, reconnectAttempts < maxReconnectAttempts else {
            return false
        }

        reconnectWorkItem?.cancel()

        let delay = reconnectBaseDelay * pow(2.0, Double(reconnectAttempts))
        let item = DispatchWorkItem { [weak self] in
            guard let self else { return }
            self.reconnectAttempts += 1

            Task {
                let success = await attempt()
                if !success {
                    self.scheduleReconnect(attempt: attempt)
                } else {
                    self.reconnectAttempts = 0
                }
            }
        }

        reconnectWorkItem = item
        DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: item)
        return true
    }

    /// Cancel any pending reconnect and reset counter.
    public func cancelReconnect() {
        reconnectWorkItem?.cancel()
        reconnectWorkItem = nil
        reconnectAttempts = 0
    }

    // MARK: - Lifecycle

    /// Called when app enters foreground — attempt session restore.
    public func appDidBecomeActive() {
        checkExpiry(onExpired: {})
    }

    deinit {
        stopExpiryMonitoring()
        cancelReconnect()
    }
}

// MARK: - Relay Health

/// Relay connection quality indicator.
public enum RelayHealth: Equatable {
    case connected
    case degraded
    case disconnected
    case reconnecting
}

File: onux/packages/android-kotlin/src/main/kotlin/com/onchainux/walletconnect/WCSessionManager.kt
/**
 * WCSessionManager — Session persistence & auto-reconnect for WalletConnect v2 (Android).
 *
 * Provides:
 * - SharedPreferences-backed session persistence (topic, accounts, chainId)
 * - Session TTL enforcement (default 7 days)
 * - Automatic reconnection with exponential backoff on app foreground
 * - Reconnect attempt limits and relay health tracking
 *
 * Used internally by WCClient; also usable standalone for custom integrations.
 */
package com.cinacoin.walletconnect

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

// ============================================================
// Persisted Session Model
// ============================================================

/** Codable snapshot of an active WC v2 session for persistence. */
data class PersistedSession(
    val topic: String,
    val accounts: List<String>,
    val chainId: Int,
    val peerName: String?,
    val createdAt: Long,
    val expiresAt: Long
) {
    /** Whether this session has exceeded its TTL. */
    val isExpired: Boolean get() = System.currentTimeMillis() >= expiresAt

    companion object {
        private const val SEP = "|"

        fun serialize(s: PersistedSession): String =
            listOf(
                s.topic,
                s.accounts.joinToString(","),
                s.chainId.toString(),
                s.peerName ?: "",
                s.createdAt.toString(),
                s.expiresAt.toString()
            ).joinToString(SEP)

        fun deserialize(raw: String): PersistedSession? = runCatching {
            val parts = raw.split(SEP)
            require(parts.size >= 6)
            PersistedSession(
                topic = parts[0],
                accounts = parts[1].takeIf { it.isNotEmpty() }?.split(",") ?: emptyList(),
                chainId = parts[2].toInt(),
                peerName = parts[3].takeIf { it.isNotEmpty() },
                createdAt = parts[4].toLong(),
                expiresAt = parts[5].toLong()
            )
        }.getOrNull()
    }
}

// ============================================================
// Relay Health
// ============================================================

/** Relay connection quality indicator. */
sealed class RelayHealth {
    object Connected : RelayHealth()
    object Degraded : RelayHealth()
    object Disconnected : RelayHealth()
    object Reconnecting : RelayHealth()
}

// ============================================================
// Session Manager
// ============================================================

/** Manages WC v2 session lifecycle: persist, restore, expiry-check, reconnect. */
class WCSessionManager(
    private val prefs: SharedPreferences,
    private val ioScope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
) {

    companion object {
        private const val KEY_SESSION = "wc_v2_persisted_session_v1"
        private const val DEFAULT_TTL_MS = 7L * 24 * 60 * 60 * 1000 // 7 days

        /** Factory: create with default SharedPreferences. */
        fun create(context: Context): WCSessionManager {
            val prefs = context.getSharedPreferences("wc_session_manager", Context.MODE_PRIVATE)
            return WCSessionManager(prefs)
        }
    }

    /** Session TTL in milliseconds (default: 7 days). */
    var sessionTTL: Long = DEFAULT_TTL_MS

    /** Whether auto-reconnect is enabled. */
    var autoReconnectEnabled: Boolean = true

    /** Maximum reconnect attempts before giving up. */
    var maxReconnectAttempts: Int = 5

    /** Base delay for exponential backoff (milliseconds). */
    var reconnectBaseDelay: Long = 2000

    /** Current reconnect attempt count. */
    var reconnectAttempts: Int = 0
        private set

    /** Relay health status. */
    private val _relayHealth = MutableStateFlow<RelayHealth>(RelayHealth.Disconnected)
    val relayHealth: StateFlow<RelayHealth> = _relayHealth.asStateFlow()

    private var reconnectJob: Job? = null
    private var expiryJob: Job? = null

    // ─── Persistence ──────────────────────────────────────────────────

    /** Save a session snapshot for later restoration. */
    fun persistSession(
        topic: String,
        accounts: List<String>,
        chainId: Int,
        peerName: String? = null
    ) {
        val session = PersistedSession(
            topic = topic,
            accounts = accounts,
            chainId = chainId,
            peerName = peerName,
            createdAt = System.currentTimeMillis(),
            expiresAt = System.currentTimeMillis() + sessionTTL
        )
        prefs.edit().putString(KEY_SESSION, PersistedSession.serialize(session)).apply()
    }

    /** Load the persisted session, if any and not expired. */
    fun loadSession(): PersistedSession? {
        val raw = prefs.getString(KEY_SESSION, null) ?: return null
        return PersistedSession.deserialize(raw)?.takeUnless { it.isExpired }
    }

    /** Clear persisted session data. */
    fun clearSession() {
        prefs.edit().remove(KEY_SESSION).apply()
    }

    // ─── Expiry Monitoring ────────────────────────────────────────────

    /** Start periodic expiry checks (every 60s). */
    fun startExpiryMonitoring(onExpired: () -> Unit) {
        expiryJob?.cancel()
        // Immediate check
        checkExpiry(onExpired)
        // Periodic check
        expiryJob = ioScope.launch {
            while (isActive) {
                delay(60_000)
                checkExpiry(onExpired)
            }
        }
    }

    /** Stop expiry monitoring. */
    fun stopExpiryMonitoring() {
        expiryJob?.cancel()
        expiryJob = null
    }

    /** Check if the persisted session has expired. */
    private fun checkExpiry(onExpired: () -> Unit) {
        val session = loadSession() ?: return
        if (session.isExpired) {
            clearSession()
            onExpired()
        }
    }

    // ─── Auto-Reconnect ───────────────────────────────────────────────

    /** Schedule an auto-reconnect with exponential backoff. Returns `true` if scheduled. */
    fun scheduleReconnect(attempt: suspend () -> Boolean): Boolean {
        if (!autoReconnectEnabled || reconnectAttempts >= maxReconnectAttempts) return false

        reconnectJob?.cancel()
        reconnectJob = ioScope.launch {
            delay(reconnectBaseDelay * (1L shl reconnectAttempts))
            reconnectAttempts++
            val success = attempt()
            if (!success) {
                scheduleReconnect(attempt)
            } else {
                reconnectAttempts = 0
            }
        }
        return true
    }

    /** Cancel any pending reconnect and reset counter. */
    fun cancelReconnect() {
        reconnectJob?.cancel()
        reconnectJob = null
        reconnectAttempts = 0
    }

    // ─── Lifecycle ────────────────────────────────────────────────────

    /** Called when app enters foreground — check expiry. */
    fun appDidBecomeActive() {
        checkExpiry {}
    }

    /** Called when app goes to background — stop monitoring. */
    fun applicationWillResignActive() {
        stopExpiryMonitoring()
    }

    /** Clean up all resources. */
    fun close() {
        stopExpiryMonitoring()
        cancelReconnect()
    }
}
